### ISSUE TITLE
Add server side calculations

### ISSUE THREAD

**Issue by charlestati** [google/boardgame.io#181]
It would allow this kind of scenario in multiplayer:

1. **Player 1** attacks **Player 2**
2. Start **Player 1** Attack animation (if `optimisticUpdate`)
3. Start **Player 2** Hurt animation (if `optimisticUpdate`)
4. Calculate damages on the server
5. Update the health state on both clients

From what I've seen, the server strips the `_random` API from `ctx` so it stays server side. Maybe it could also strip any `flow.secret.*` so we could use `flow.secret.onMove` for moves calculated server side?

If you have any hints for where to start I could dig into it!

**Comment by nicolodavis** [comment]
[onMove](https://github.com/google/boardgame.io/blob/master/src/core/flow.js#L171) is already calculated on the server side. Does this solve your use-case?

**Comment by charlestati** [comment]
Yes, but it is still visible client-side, right?
It would be an issue if, for some reason, the calculation must be kept secret.

Also, it works for simple moves but for games with hundreds of different moves (such as role-playing games) each player would download lots of unnecessary data.

**Comment by nicolodavis** [comment]
The framework currently does not allow any secret logic, just secret data. Maybe we should just add an option to strip away everything in the `flow` object on the client side (at the expensive of optimistic updates). That way you can just use `onMove` to do secret computations and they aren't visible on the client side.

**Comment by charlestati** [comment]
Yes, when `optimisticUpdate` is disabled stripping away everything in the `flow` could be a solution!

Thank you for your help!

**Comment by Rupesh-ark** [comment]
This is supported now: declare the move in long form with `client: false` and it executes only on the server, with `redact: true` to also hide its payload from the log — see the moves section of https://boardgame.io/documentation/#/api/Game. Closing as fixed.
